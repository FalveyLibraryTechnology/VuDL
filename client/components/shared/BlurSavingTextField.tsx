import React, { useState } from "react";
import TextField from "@mui/material/TextField";

export interface BlurSavingTextFieldProps {
    value: string;
    setValue: (value: string) => void;
}

const BlurSavingTextField = ({ value, setValue }: BlurSavingTextFieldProps): React.ReactElement => {
    const [temporaryValue, setTemporaryValue] = useState(value);
    const handleBlurEvent = (event: React.ChangeEvent) => {
        setValue(event.target.value);
    };
    const handleChangeEvent = (event: React.ChangeEvent) => {
        setTemporaryValue(event.target.value);
    };
    return <TextField value={temporaryValue} onChange={handleChangeEvent} onBlur={handleBlurEvent} />;
};

export default BlurSavingTextField;
