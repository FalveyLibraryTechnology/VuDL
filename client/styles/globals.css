:root,
[color-scheme="light"] {
    --body-bg: #fafbfc;
    --heading-color: #4d4d4d;
    --link-color: #008aed;

    --btn-bg: #485a6c;
    --btn-text: #ffffff;
    --btn-border: #275575;
    --btn-focus-bg: #3498db;

    --btn-primary-bg: #27ae60;
    --btn-primary-text: #ffffff;
    --btn-primary-border: #275575;
    --btn-primary-focus-bg: #155724;

    --btn-danger-bg: red;
    --btn-danger-text: #ffffff;
    --btn-danger-border: #a10705;
    --btn-danger-focus-bg: #a10705;

    --btn-logout-bg: #1abc9c;
    --btn-logout-text: #ffffff;
    --btn-logout-border: #16a085;
    --btn-logout-focus-bg: #1abc9c;

    --btn-toggle-bg: #7f8c8d;
    --btn-toggle-text: #ffffff;
    --btn-toggle-border: #34495e;
    --btn-toggle-focus-bg: #53626f;

    --group-border: #53626f;

    --input-border: #6a737d;
    --input-focus-border: #2188ff;

    --label-hover: #c8e1ff;

    --page-list-bg: #ffffff;
    --page-list-border: #c1c1c1;

    --thumbnail-bg: #e1e1e1;
    --thumbnail-focus-bg: #51aded;
    --thumbnail-number-bg: #ffffff;
    --thumbnail-number-text: #000000;
    --thumbnail-preview-border: #000000;

    --tree-item-focus-bg: #000000;
    --tree-item-focus-text: #ffffff;

    --shadow-gray: #c8e1ff;
}

[color-scheme="dark"] {
}
