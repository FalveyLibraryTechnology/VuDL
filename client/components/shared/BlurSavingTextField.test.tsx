import React from "react";
import { describe, beforeEach, expect, it } from "@jest/globals";
import renderer from "react-test-renderer";
import BlurSavingTextField, { BlurSavingTextFieldProps } from "./BlurSavingTextField";

jest.mock("@mui/material/TextField", function () {
    return (props) => "TextField: " + JSON.stringify(props);
});

describe("BlurSavingTextField", () => {
    let props: BlurSavingTextFieldProps;
    let value: string;
    let setterWasCalled: boolean;
    beforeEach(() => {
        value = "foo";
        setterWasCalled = false;
        props = {
            value,
            setValue: (x: string) => {
                setterWasCalled = true;
                value = x;
            },
        };
    });

    it("renders using the initial value", () => {
        const tree = renderer.create(<BlurSavingTextField {...props} />).toJSON();
        expect(tree).toMatchSnapshot();
    });

    it("updates the temporary internal value appropriately", () => {
        const tree = renderer.create(<BlurSavingTextField {...props} />);
        renderer.act(() => {
            tree.root.children[0].props.onChange({ target: { value: "bar" } });
        });
        expect(tree.toJSON()).toMatchSnapshot();
        // Even though the internal value has changed, the external hasn't (because
        // we didn't trigger blur yet):
        expect(value).toEqual("foo");
    });

    it("saves on blur", () => {
        const tree = renderer.create(<BlurSavingTextField {...props} />);
        renderer.act(() => {
            tree.root.children[0].props.onBlur({ target: { value: "bar" } });
        });
        expect(setterWasCalled).toEqual(true);
        expect(value).toEqual("bar");
    });

    it("doesn't call the save callback if nothing has changed", () => {
        const tree = renderer.create(<BlurSavingTextField {...props} />);
        renderer.act(() => {
            tree.root.children[0].props.onBlur({ target: { value: "foo" } });
        });
        expect(setterWasCalled).toEqual(false);
        expect(value).toEqual("foo");
    });
});
