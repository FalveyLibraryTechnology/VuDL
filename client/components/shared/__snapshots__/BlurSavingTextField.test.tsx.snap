// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BlurSavingTextField renders using the initial value 1`] = `"TextField: {"value":"foo"}"`;

exports[`BlurSavingTextField updates the temporary internal value appropriately 1`] = `"TextField: {"value":"bar"}"`;
